File: logica_computacional_II/LogicaPredicadosSimples.java
import java.util.HashMap; // Importa a classe HashMap para criar mapas (chave-valor).
import java.util.Map;     // Importa a interface Map que representa um mapa.
import java.util.Set;     // Importa a interface Set que representa um conjunto de elementos únicos.

// Representa um objeto no nosso domínio (o universo de discurso).
class Objeto {
    String nome; // O nome do objeto.

    // Construtor da classe Objeto.
    public Objeto(String nome) {
        this.nome = nome;
    }

    // Sobrescreve o método toString() para uma representação textual amigável do objeto.
    @Override
    public String toString() {
        return nome;
    }

    // Sobrescreve o método equals() para comparar objetos pelo seu nome.
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true; // Se for o mesmo objeto na memória.
        if (obj == null || getClass() != obj.getClass()) return false; // Se for nulo ou de outra classe.
        Objeto objeto = (Objeto) obj; // Faz um cast para o tipo Objeto.
        return nome.equals(objeto.nome); // Compara os nomes dos objetos.
    }

    // Sobrescreve o método hashCode() para ser consistente com o método equals().
    @Override
    public int hashCode() {
        return nome.hashCode(); // Usa o hashCode da String nome.
    }
}

// Interface para representar um predicado que aceita um objeto como argumento e retorna um booleano.
interface PredicadoUnario {
    boolean avaliar(Objeto x); // Método a ser implementado para avaliar o predicado para um objeto.
}

// Interface para representar um predicado que aceita dois objetos como argumentos e retorna um booleano.
interface PredicadoBinario {
    boolean avaliar(Objeto x, Objeto y); // Método a ser implementado para avaliar o predicado para dois objetos.
}

public class LogicaPredicadosSimples {

    public static void main(String[] args) {
        // Definindo nosso domínio de objetos (as entidades sobre as quais falaremos).
        Set<Objeto> dominio = Set.of(
                new Objeto("alice"),
                new Objeto("bob"),
                new Objeto("carol")
        );

        // Definindo alguns predicados unários (propriedades de um único objeto).
        Map<String, PredicadoUnario> predicadosUnarios = new HashMap<>();
        // O predicado "EhHumano" é verdadeiro para "alice" e "bob".
        predicadosUnarios.put("EhHumano", x -> x.nome.equals("alice") || x.nome.equals("bob"));
        // O predicado "EhMulher" é verdadeiro para "alice" e "carol".
        predicadosUnarios.put("EhMulher", x -> x.nome.equals("alice") || x.nome.equals("carol"));

        // Definindo alguns predicados binários (relações entre dois objetos).
        Map<String, PredicadoBinario> predicadosBinarios = new HashMap<>();
        // O predicado "Ama" é verdadeiro se "alice" ama "bob" ou "bob" ama "alice".
        predicadosBinarios.put("Ama", (x, y) -> (x.nome.equals("alice") && y.nome.equals("bob")) ||
                                            (x.nome.equals("bob") && y.nome.equals("alice")));

        // Avaliando algumas expressões simples da lógica de predicados.

        // Criando instâncias de objetos do domínio para usar nas avaliações.
        Objeto alice = new Objeto("alice");
        Objeto carol = new Objeto("carol");
        Objeto bob = new Objeto("bob");

        // Avaliando um predicado unário para um objeto específico.
        System.out.println("EhHumano(alice): " + predicadosUnarios.get("EhHumano").avaliar(alice)); // Saída: true
        System.out.println("EhHumano(carol): " + predicadosUnarios.get("EhHumano").avaliar(carol)); // Saída: false

        // Avaliando um predicado binário para dois objetos específicos.
        System.out.println("Ama(alice, bob): " + predicadosBinarios.get("Ama").avaliar(alice, bob)); // Saída: true
        System.out.println("Ama(bob, carol): " + predicadosBinarios.get("Ama").avaliar(bob, carol)); // Saída: false

        // Simulação rudimentar do quantificador existencial (para um domínio finito).
        // Verifica se existe algum objeto no domínio para o qual o predicado "EhHumano" é verdadeiro.
        boolean existeHumano = false;
        for (Objeto obj : dominio) {
            if (predicadosUnarios.get("EhHumano").avaliar(obj)) {
                existeHumano = true;
                break; // Se encontrarmos um humano, não precisamos verificar os outros.
            }
        }
        System.out.println("Existe x tal que EhHumano(x): " + existeHumano); // Saída: true

        // Simulação rudimentar do quantificador universal (para um domínio finito).
        // Verifica se o predicado "EhHumano" é verdadeiro para todos os objetos no domínio.
        boolean todosSaoHumanos = true;
        for (Objeto obj : dominio) {
            if (!predicadosUnarios.get("EhHumano").avaliar(obj)) {
                todosSaoHumanos = false;
                break; // Se encontrarmos um não humano, não precisamos verificar os outros.
            }
        }
        System.out.println("Para todo x, EhHumano(x): " + todosSaoHumanos); // Saída: false

        // Combinação da avaliação de dois predicados usando um conectivo lógico (AND - && em Java).
        // Verifica se "alice" é tanto humana quanto mulher.
        boolean aliceEhHumanoEMulher = predicadosUnarios.get("EhHumano").avaliar(alice) &&
                                        predicadosUnarios.get("EhMulher").avaliar(alice);
        System.out.println("EhHumano(alice) E EhMulher(alice): " + aliceEhHumanoEMulher); // Saída: true
    }
}

File: logica_computacional_II/LogicaPredicadosSimples.py
# Representa um objeto no nosso domínio.
class Objeto:
    def __init__(self, nome):
        self.nome = nome

    def __str__(self):
        return self.nome

    def __repr__(self):
        return f"Objeto('{self.nome}')"

    def __eq__(self, other):
        if isinstance(other, Objeto):
            return self.nome == other.nome
        return NotImplemented

    def __hash__(self):
        return hash(self.nome)

# Representa um predicado com um argumento.
def predicado_unario(nome_predicado, dominio, avaliacao):
    """
    Cria um predicado unário.

    Args:
        nome_predicado (str): O nome do predicado.
        dominio (set): O conjunto de objetos no domínio.
        avaliacao (callable): Uma função que recebe um objeto e retorna True ou False.

    Returns:
        callable: Uma função que recebe um objeto e retorna o valor de verdade do predicado para esse objeto.
    """
    def avaliar_predicado(x):
        if x in dominio:
            return avaliacao(x)
        return False  # Assumindo que o predicado é falso para objetos fora do domínio

    avaliar_predicado.nome = nome_predicado
    return avaliar_predicado

# Representa um predicado com dois argumentos.
def predicado_binario(nome_predicado, dominio, avaliacao):
    """
    Cria um predicado binário.

    Args:
        nome_predicado (str): O nome do predicado.
        dominio (set): O conjunto de objetos no domínio.
        avaliacao (callable): Uma função que recebe dois objetos e retorna True ou False.

    Returns:
        callable: Uma função que recebe dois objetos e retorna o valor de verdade do predicado para esses objetos.
    """
    def avaliar_predicado(x, y):
        if x in dominio and y in dominio:
            return avaliacao(x, y)
        return False  # Assumindo que o predicado é falso para objetos fora do domínio

    avaliar_predicado.nome = nome_predicado
    return avaliar_predicado

if __name__ == "__main__":
    # Definindo nosso domínio de objetos.
    dominio = {Objeto("alice"), Objeto("bob"), Objeto("carol")}

    # Definindo alguns predicados sobre nosso domínio.
    eh_humano = predicado_unario("EhHumano", dominio, lambda x: x.nome == "alice" or x.nome == "bob")
    eh_mulher = predicado_unario("EhMulher", dominio, lambda x: x.nome == "alice" or x.nome == "carol")
    ama = predicado_binario("Ama", dominio, lambda x, y: (x.nome == "alice" and y.nome == "bob") or
                                                      (x.nome == "bob" and y.nome == "alice"))

    # Criando instâncias de objetos do domínio para usar nas avaliações.
    alice = Objeto("alice")
    bob = Objeto("bob")
    carol = Objeto("carol")

    # Avaliando predicados para objetos específicos.
    print(f"{eh_humano.nome}(alice): {eh_humano(alice)}")  # Saída: True
    print(f"{eh_humano.nome}(carol): {eh_humano(carol)}")  # Saída: False
    print(f"{ama.nome}(alice, bob): {ama(alice, bob)}")    # Saída: True
    print(f"{ama.nome}(bob, carol): {ama(bob, carol)}")    # Saída: False

    # Simulação rudimentar de quantificador existencial (para um domínio finito).
    existe_humano = any(eh_humano(x) for x in dominio)
    print(f"Existe x tal que {eh_humano.nome}(x): {existe_humano}")  # Saída: True

    # Simulação rudimentar de quantificador universal (para um domínio finito).
    todos_sao_humanos = all(eh_humano(x) for x in dominio)
    print(f"Para todo x, {eh_humano.nome}(x): {todos_sao_humanos}")  # Saída: False

    # Combinação com conectivo lógico (AND).
    alice_eh_humano_e_mulher = eh_humano(alice) and eh_mulher(alice)
    print(f"{eh_humano.nome}(alice) and {eh_mulher.nome}(alice): {alice_eh_humano_e_mulher}")  # Saída: True
